### PR TITLE
TextIO/CompressedSource: support splitting AUTO mode files into bundles

### PR DESCRIPTION
**Pull request by dhalperi**
R: @jkff please take a look. This code has been the cause of major bugs in the past.

I am still planning a pass tomorrow to clean up the tests -- so much redundant code, but it's so entangled it's a pain.

### CODE REVIEW

**Comment by dhalperi** [comment]
Revamped the tests on the airplane; see commit comments for thoughts and also suggestions for further improvement are welcome!

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 1]
```diff
@@ -108,19 +120,93 @@
...
+  }
+
   @BeforeClass
-  public static void setupClass() {
+  public static void setupClass() throws IOException {
     IOChannelUtils.registerStandardIOFactories(TestPipeline.testingPipelineOptions());
+    tempFolder =  Files.createTempDirectory("TextIOTest");
+    // empty files
+    emptyTxt = writeToFile(EMPTY, "empty.txt", CompressionType.UNCOMPRESSED);
+    emptyGz = writeToFile(EMPTY, "empty.gz", GZIP);
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Perhaps one more thing to test against: concatenated gzip files (https://issues.apache.org/jira/browse/BEAM-167)

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 2]
```diff
@@ -748,87 +699,85 @@ private String createZipFile(List<String> expected, @Nullable String filename, S
...
-   * We expect a PCollection with the lines.
-   */
   @Test
   @Category(NeedsRunner.class)
-  public void testZipCompressedRead() throws Exception {
-    String[] lines = {"Irritable eagle", "Optimistic jay", "Fanciful hawk"};
-    List<String> expected = new ArrayList<>();
-
-    String filename = createZipFile(expected, null, lines);
+  public void testTxtReadRead() throws Exception {
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Strange name: why "ReadRead"?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 3]
```diff
@@ -748,87 +699,85 @@ private String createZipFile(List<String> expected, @Nullable String filename, S
...
+      assertReadingCompressedFileMatchesExpected(emptyTxt, type, EMPTY);
+      assertReadingCompressedFileMatchesExpected(tinyTxt, type, TINY);
+      assertReadingCompressedFileMatchesExpected(largeTxt, type, LARGE);
+    }
+  }
 
-    Pipeline p = TestPipeline.create();
+  @Test
+  @Category(NeedsRunner.class)
+  public void testGzipCompressedRead() throws Exception {
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Is it possible to further reduce code duplication between testing different compression algorithms?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 4]
```diff
@@ -839,16 +788,9 @@ public void testZipCompressedReadWithMultiEntriesFile() throws Exception {
...
-
-    Pipeline p = TestPipeline.create();
-
-    TextIO.Read.Bound<String> read =
-        TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP);
-    PCollection<String> output = p.apply(read);
-
-    PAssert.that(output).containsInAnyOrder(expected);
-    p.run();
+    String filename = createZipFile(expected, "multiple entries", entry0, entry1, entry2);
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Add a comment clarifying whether we should read the first entry or all entries.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 5]
```diff
@@ -1116,12 +1051,79 @@ public void testSplittingSourceWithMixedDelimitersAndNonEmptyBytesAtEnd() throws
...
+    // Sanity check: file is at least 2 bundles long.
+    assertThat(largeTxt.length(), greaterThan(2 * desiredBundleSize));
+
+    FileBasedSource<String> source = TextIO.Read.from(largeTxt.getPath()).getSource();
+    List<? extends FileBasedSource<String>> splits =
+        source.splitIntoBundles(desiredBundleSize, options);
+
+    // At least 2 splits and they are equal to reading the whole file.
+    assertThat(splits, hasSize(greaterThan(1)));
+    SourceTestUtils.assertSourcesEqualReferenceSource(source, splits, options);
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: import static?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 6]
```diff
@@ -1116,12 +1051,79 @@ public void testSplittingSourceWithMixedDelimitersAndNonEmptyBytesAtEnd() throws
...
+  }
+
+ @Test
+  public void testInitialSplitIntoBundlesAutoModeGz() throws Exception {
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Deduplicate between compression algorithms?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java (hunk 2 above)
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Fixed -- some weird replacement typo.

**Comment by dhalperi** [comment]
I have a separate further-test-revamping branch, but I'm still not happy with it. Brainstorming welcome!

https://github.com/dhalperi/incubator-beam/commit/969615890d471d61187da0d83c38a282baabaa55

On the other hand, I'd like to unblock this relatively quickly b/c there's some time pressure with respect to Dataflow.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java (hunk 1 above)
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Tested: https://github.com/apache/incubator-beam/pull/114/files#diff-19e64d896b88de9a6fcdf9a7d2867df6R126  (but also added in the next test refactoring branch)

**Comment by jkff** [comment]
If this is blocking something, I'm fine getting it in as-is and refactoring the tests afterwards. So, LGTM.

I looked at the test-revamping branch and the use of @Parameterized seems reasonable.

**Comment by dhalperi** [comment]
Thanks

**Pull request closed without merging**